### DISCUSSION TITLE
How to check a particular error was returned by a previous step?

### DISCUSSION THREAD

**Discussion by will-hart** [cucumber-rs/cucumber#329] [Q&A] [answered]
In some cases I want to assert that a particular error was returned when I perform a step. For instance in a toy example I would love to have:

```gherkin
Scenario: Divide by n
    Given I start with 6
    When I divide by 6
    Then I have 1
Scenario: Divide by 0
    Given I start with 6
    When I divide by 0
    Then I get a "divide by 0" error
```

How would I implement an expected failure? Ideally the `When I divide by (\d+)` step would be the same for both scenarios, but if I return an `Err(_)` result from the `When` step then the test fails.

Should I be caching the result of the `When` step in the `World`? e.g.

```rust
pub struct MySpecialInteger(u64);

#[derive(Debug, Default, World)]
pub struct MyWorld{
    number: MySpecialInteger,
    result: Result<u64, MyNumberError>
}

#[when(r"I divide by (\d+)"])
fn divide_by_n(world: &mut MyWorld, num: u64) {
    world.result = world.number.0 / num; // pretend this produces a Result
}

// pseudocode
#[then(r"I get a (\s+) error"])
fn divide_by_n(world: &mut MyWorld, expected_message: &str) {
    assert!(world.result.message.contains(expected_message));
}
```

I could also have a then step which is `Then dividing by (\d+) produces error xyz`, but this seems like mixing steps and assertions.

**Comment by ilslv** [comment, marked as answer]
> I could also have a then step which is `Then dividing by (\d+) produces error xyz`, but this seems like mixing steps and assertions.

You are absolutely right, the recommended way is to store whole `Result`s on `when` steps and assert them on `then`s.
